### ISSUE TITLE
No error when typedef is missing semicolon

### ISSUE THREAD

**Issue by tobil4sk** [HaxeFoundation/haxe#10822]
This is permitted in Haxe:
```haxe
typedef A = String
```
I think it is a bug to do with allowing:
```haxe
typedef A = { a: String }
```

The code seems to be found here: https://github.com/HaxeFoundation/haxe/blob/development/src/syntax/grammar.mly#L284-L286

We should instead do something like we do here:
https://github.com/HaxeFoundation/haxe/blob/development/src/syntax/grammar.mly#L100

This is inconsistent with the rest of the language, and causes issues with the vshaxe parser:
https://github.com/vshaxe/vshaxe/issues/24
https://github.com/vshaxe/haxe-TmLanguage/issues/54

**Comment by Simn** [comment]
I checked the blame because I thought this was something broken more recently, but it's from 2007: https://github.com/HaxeFoundation/haxe/commit/e7aae26decad1dab736b77df6a6790c2c1642c89

I agree though, omitting the semicolon here is wrong because it's supposed to be optional only after `}`. My concern is that this might break some ancient libraries that nobody really maintains anymore, but I suppose there's only one way to find out.

**Comment by Simn** [comment]
Well, turns out there are several places in our standard library and tests already, which doesn't surprise me.

IMO it's better to just accept this and support it in vshaxe. Fixing this might technically be "correct", but I don't really see how it makes anyone's Haxe life any easier.

Edit: I've pushed the branch so that we can gauge the fallout.

**Comment by tobil4sk** [comment]
I guess at least it would be a pretty unambiguous error message. I just think that if we're going to have optional semicolons in some places then the rules should be consistent.

Looks like it does break tink though, so maybe it is a bit problematic, however, it is a really simple fix. Maybe it could be considered for Haxe 5? If any of the other parsing issues end up being resolved it would be nice to have a general parser cleanup for that release, because right now inconsistency like this makes writing external parsers annoying, especially with tools that are more strict about grammar correctness. I guess this case isn't too bad though, because as far as I can tell it shouldn't cause ambiguity.

**Comment by Simn** [comment]
Yeah I don't think we should change this in Haxe 4.
